master/internal/stream/test_util.go

```diff
@@ -120,135 +144,117 @@ func (s *mockSocket) Close() {
-		switch typedMsg := msg.(type) {
-		case stream.DeleteMsg:
-			deletions = append(deletions, typedMsg.Deleted)
-		case stream.UpsertMsg:
-			upsertM, ok := typedMsg.Msg.(M)
-			if !ok {
-				t.Errorf("%s: expected %T, but received %T", testCaseDescription, typeHolder, typedMsg.Msg)
-			}
-			upserts = append(upserts, upsertM)
-		case SyncMsg:
-			syncs = append(syncs, typedMsg)
+		switch {
+		case deletePattern.MatchString(msg):
+			deletions = append(deletions, msg)
+		case upsertPattern.MatchString(msg):
+			upserts = append(upserts, msg)
+		case syncPattern.MatchString(msg):
+			syncs = append(syncs, msg)
 		default:
-			t.Errorf("%s: expected a string or %T, but received %T",
-				testCaseDescription,
-				typeHolder,
-				reflect.TypeOf(msg).Name(),
-			)
+			t.Errorf("unknown message type: %s", msg)
 		}
 	}
 	return deletions, upserts, syncs
 }
 
-func validateMsgs[M stream.Msg](
+func listToSet(l []string) map[string]struct{} {
+	set := make(map[string]struct{}, len(l))
+	for _, val := range l {
+		set[val] = struct{}{}
+	}
+	return set
+}
+
+func validateMsgs(
 	t *testing.T,
-	testCaseDescription string,
-	sync SyncMsg,
-	expectedSync SyncMsg,
-	upserts []stream.Msg,
-	expectedUpserts []M,
+	sync string,
+	expectedSync string,
+	upserts []string,
+	expectedUpserts []string,
 	deletions []string,
 	expectedDeletions []string,
 ) {
+	expectedUpsertSet := listToSet(expectedUpserts)
+	expectedDeletionSet := listToSet(expectedDeletions)
+
 	switch {
 	// check if we received the correct number of trial messages
 	case len(upserts) != len(expectedUpserts):
 		t.Errorf(
-			"%s: did not receive expected number of upsert messages: expected %d, actual: %d",
-			testCaseDescription,
+			"did not receive expected number of upsert messages:\n\texpected %d\n\tactual: %d",
 			len(expectedUpserts),
 			len(upserts),
 		)
 	// check if we received the correct number of deletion messages
 	case len(deletions) != len(expectedDeletions):
 		t.Errorf(
-			"%s: did not receive expected number of deletion messages: expected %v, actual: %v",
-			testCaseDescription,
+			"did not receive expected number of deletion messages:\n\texpected %v\n\tactual: %v",
 			len(expectedDeletions),
 			len(deletions),
 		)
 	// check if we receieved the correct SyncMsg
-	case sync.SyncID != expectedSync.SyncID:
+	case sync != expectedSync:
 		t.Errorf(
-			"%s: did not receive expected sync message: expected: %v, actual: %v",
-			testCaseDescription,
+			"did not receive expected sync message:\n\texpected: %v\n\tactual: %v",
 			expectedSync,
 			sync,
 		)
 	// check if content of messages is correct
 	default:
-		// XXX: this expects messages to be sent in a deterministic order, is this actually enforced?
-		// should msgs be sorted then?
 		for i := range upserts {
-			if err := validateUpsertMsg(upserts[i], expectedUpserts[i]); err != nil {
-				t.Errorf("%s: %s", testCaseDescription, err.Error())
+			if _, ok := expectedUpsertSet[upserts[i]]; !ok {
```

Comment:
I'm thinking maybe it's better to reverse this - so putting received upserts in a map and looping over the expected. Reason being if we happen to receive duplicate upsert messages, we won't catch it. If it were the other way around, we would make sure we loop over every one of the messages we want.

Reply:
Good call, I had originally thought we could get around this by having the len check in the beginning; but we do run the risk of missing duplicate messages!